typescript/bots/FunLib/aba_defend.ts

```diff
@@ -0,0 +1,382 @@
+import * as jmz from "bots/FunLib/jmz_func";
+import {
+    BotActionDesire,
+    BotMode,
+    BotModeDesire,
+    BotScriptEnums,
+    Lane,
+    Unit,
+    UnitType,
+    Vector,
+} from "bots/ts_libs/dota";
+import {
+    GameStates,
+    IsPingedByAnyPlayer,
+    IsValidHero,
+} from "bots/FunLib/utils";
+import { add } from "bots/ts_libs/utils/native-operators";
+
+const furthestBuildings = {
+    [Lane.Top]: {
+        towers: [
+            { id: BotScriptEnums.TOWER_TOP_1, mulMax: 0.5, mulMin: 1 },
+            { id: BotScriptEnums.TOWER_TOP_2, mulMax: 1, mulMin: 2 },
+            { id: BotScriptEnums.TOWER_TOP_3, mulMax: 1.5, mulMin: 2 },
+        ],
+        barracks: [
+            BotScriptEnums.BARRACKS_TOP_MELEE,
+            BotScriptEnums.BARRACKS_TOP_RANGED,
+        ],
+    },
+    [Lane.Bot]: {
+        towers: [
+            { id: BotScriptEnums.TOWER_BOT_1, mulMax: 0.5, mulMin: 1 },
+            { id: BotScriptEnums.TOWER_BOT_2, mulMax: 1, mulMin: 2 },
+            { id: BotScriptEnums.TOWER_BOT_3, mulMax: 1.5, mulMin: 2 },
+        ],
+        barracks: [
+            BotScriptEnums.BARRACKS_BOT_MELEE,
+            BotScriptEnums.BARRACKS_BOT_RANGED,
+        ],
+    },
+    [Lane.Mid]: {
+        towers: [
+            { id: BotScriptEnums.TOWER_MID_1, mulMax: 0.5, mulMin: 1 },
+            { id: BotScriptEnums.TOWER_MID_2, mulMax: 1, mulMin: 2 },
+            { id: BotScriptEnums.TOWER_MID_3, mulMax: 1.5, mulMin: 2 },
```

Comment:
we agreed on not using this enum outside of the enums file. let's make new enums for Tower